prow/inrepoconfig/api/api.go

```diff
@@ -0,0 +1,151 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"github.com/sirupsen/logrus"
+	"sigs.k8s.io/yaml"
+
+	checkconfigapi "k8s.io/test-infra/prow/cmd/checkconfig/api"
+	"k8s.io/test-infra/prow/config"
+	"k8s.io/test-infra/prow/git"
+	"k8s.io/test-infra/prow/github"
+)
+
+const (
+	// PluginName is the name of this plugin
+	PluginName = "inrepoconfig"
+
+	// ContextName is the name of the context the inrepoconfig plugin creates at GitHub
+	ContextName = "prow-config-parsing"
+
+	// ConfigFileName is the name of the configfile the inrepoconfig plugin uses to read its
+	// config from
+	ConfigFileName = "prow.yaml"
+)
+
+// InRepoConfig is the type `prow.yaml` gets deserialized into
+type InRepoConfig struct {
+	Presubmits []config.Presubmit `json:"presubmits,omitempty"`
+}
+
+// defaultAndValidateInRepoConfig defaults and validates the InRepoConfig. This must be called before
+// accessing any data in it
+func (irc *InRepoConfig) defaultAndValidateInRepoConfig(c *config.Config, orgRepo string) error {
+	for i := range irc.Presubmits {
+		ps := &irc.Presubmits[i]
+		config.DefaultPresubmitFields(&c.ProwConfig, ps)
+		if ps.Decorate {
+			ps.DecorationConfig = ps.DecorationConfig.ApplyDefault(c.Plank.DefaultDecorationConfig)
+		}
+		if err := config.ResolvePresets(
+			ps.Name, ps.Labels, ps.Spec, ps.BuildSpec, c.Presets); err != nil {
+			return err
+		}
+	}
+
+	// Has to be done in the end, otherwise the jobs Trigger field is unset
+	if err := config.SetPresubmitRegexes(irc.Presubmits); err != nil {
+		return fmt.Errorf("failed to set Presubmit regexes: %v", err)
+	}
+
+	for _, presubmit := range irc.Presubmits {
+		if err := checkconfigapi.ValidatePresubmitJob(orgRepo, presubmit); err != nil {
+			return err
+		}
+	}
+	ap := &additionalPresubmits{
+		fullRepoName: orgRepo,
+		presubmits:   irc.Presubmits,
+	}
+	// This is not extremely efficient as it validates all Job types on all Repos of this
+	// Prow instance.
+	// Changing that requires a bigger refactoring of config.Config.ValidateJobConfig thought,
+	// in order to keep the change required when introducing `inrepoconfig` optimizing this is
+	// kept as a future task.
+	// TODO @alvaroaleman: Revisit and improve this
+	if err := c.ValidateJobConfig(ap); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// New returns an InRepoConfig for the passed in refs
+func New(log *logrus.Entry, c *config.Config, gc *git.Client, org, repo, baseRef string, headRefs []string) (*InRepoConfig, error) {
+	mergeStrategyRaw := ""
+	githubMergeMethod := c.Tide.MergeMethod(org, repo)
+	switch githubMergeMethod {
+	case github.MergeRebase:
+		return nil, fmt.Errorf("merge strategy %s is currently not supported by %s", githubMergeMethod, PluginName)
```

Comment:
Might be nice to add to `checkconfig`

Reply:
What for? The configuration for the MergeMethod still comes from the main config. Defaulting and Validation of the config that is read from `prow.yaml` happens inside `defaultAndValidateInRepoConfig`

Reply:
To ensure someone isn't trying to turn on these plugins for a repo they have configured to rebase with tide, since it will fall over here

Reply:
Ah thats what you mean, yeah, makes perfect sense, will do